pkg/server/gitproviders/gitprovider.go

```diff
@@ -33,6 +33,7 @@ func (s *GitProviderService) GetGitProviderForUrl(repoUrl string) (gitprovider.G
 			if err == nil {
 				return gitProvider, p.Id, nil
 			}
+			return nil, "", err
```

Comment:
Why did you add this line?

There was not return here on purpose because, for example, you can have 2 GitHub accounts linked and attempt to create a workspace from a private repo on the second account. Both the providers will return true for `CanHandle` because the source is github.com but the first account won't be able to get the context because the repo is private. If we return the error here, this scenario will never be possible.

Reply:
I see. I added this because, if there's an error getting the context at L30, it wasn’t being displayed to the user, and instead, `cannot get public client for the URL` was shown. 
Can you tell me what should be the best way to handle this?

Reply:
What's the output of `dtn gp ls -f json`?

Reply:
```json
[
  {
    "Id": "a9907c79074d",
    "ProviderId": "gogs",
    "Name": "Gogs",
    "Username": "test",
    "BaseApiUrl": "https://6b93-160-22-60-16.ngrok-free.app",
    "Token": "",
    "Alias": "test"
  }
]
```

Reply:
That seems fine. What's the stacktrace of that error that happens? (Which functions actually makes the error happen in the end and who called that function)

Reply:
In `GetGitProviderForUrl` fn, the first loop doesn't return because `GetRepositoryContext` throws an error (as I am returning an error from `GetPrContext`). Then, in the second loop, g.BaseApiUrl is always empty, so canHandle return false, which leads to "cannot get public client for the URL" error.


```go
func (s *GitProviderService) GetGitProviderForUrl(repoUrl string) (gitprovider.GitProvider, string, error) {
	gitProviders, err := s.configStore.List()
	if err != nil {
		return nil, "", err
	}

	for _, p := range gitProviders {
		gitProvider, err := s.GetGitProvider(p.Id)
		if err != nil {
			continue
		}

		canHandle, _ := gitProvider.CanHandle(repoUrl)
		if canHandle {
			_, err = gitProvider.GetRepositoryContext(gitprovider.GetRepositoryContext{
				Url: repoUrl,
			})
			if err == nil {
				return gitProvider, p.Id, nil
			}
			// return nil, "", err
		}
	}

	for _, p := range config.GetSupportedGitProviders() {
		gitProvider, err := s.newGitProvider(&gitprovider.GitProviderConfig{
			ProviderId: p.Id,
			Id:         p.Id,
			Username:   "",
			Token:      "",
			BaseApiUrl: nil,
		})
		if err != nil {
			continue
		}
		canHandle, _ := gitProvider.CanHandle(repoUrl)
		if canHandle {
			return gitProvider, p.Id, nil
		}
	}

	return nil, "", errors.New("can not get public client for the URL " + repoUrl)
}
```

Reply:
@tarunrajput ah, this happens because gogs throws an error for getting the PR context. I get it now.

We need to work through that. Adding `return nil, "", err` is not acceptable because of the scenario I described above.

Since this is something specific to the gogs provider, we should add some kind of exception there.

Reply:
I believe this is not specific to gogs, I can replicate the same behavior with gitea, azure as well. If any of these fns - `ParseStaticGitContext, GetPrContext, GetBranchByCommit, GetDefaultBranch, or GetLastCommitSha` returns an error, we receive "cannot get public client for the URL" instead of the actual error.

Reply:
That's actually ok because if none of your git providers can fetch the repository context, then that means that you can't clone the repo with any of them and the function should fail.

Maybe I'm missing something. Do you have a scenario that should work but doesn't with this current setup?

Reply:
@tarunrajput I see the changes. Can you let me know which scenario didn't work before but does with the changes?

Reply:
Previously, the only issue was that I couldn't propagate the GetPrContext error to the user. With these changes, we now check if none of the providers can get the repo context, but if at least one og them can handle the repoUrl, we return the gitprovider instead of nil. This allows us to display the original error. 
I'm not certain if this is the ideal approach, but it's the best solution I could come up with.

Reply:
I understand.

I believe the simplest approach would be to gather all those that `CanHandle`. If the list is not empty, then simply return the first from the list before going through the supported git providers.

Reply:
done